src/ORM/ArrayList.php

```diff
@@ -175,23 +175,29 @@ public function toNestedArray()
         return $result;
     }
 
-    /**
-     * Get a sub-range of this dataobjectset as an array
-     * Pass null to "remove the limit" - this is for consistency with DataList::limit(null) which itself will
-     * call SQLSelect::setLimit(null)
-     *
-     * @param int|null $length
-     * @param int $offset
-     * @return static
-     */
-    public function limit($length, $offset = 0)
+    public function limit(?int $length, int $offset = 0): static
     {
-        if (!$length) {
-            $length = count($this->items ?? []);
+        if ($length === null) {
+            // If we unset the limit, we set the length to the size of the list. We still want the offset to be picked up
+            $length = sizeof($this->items);
+        }
+
+        if ($length < 0) {
+            throw new InvalidArgumentException("\$length can not be negative. $length was provided.");
+        }
+
+        if ($offset < 0) {
+            throw new InvalidArgumentException("\$offset can not be negative. $offset was provided.");
         }
```

Comment:
We weren't validating for this before. Arguably if you give a negative length, methods like array slice will want to start from the end.

While this might make sense for an array most DB don't think like that. Seems more logical to throw an exception.